### PR TITLE
[PINOT-4885] Changing SingleValue Forward index to be expandable

### PR DESCRIPTION
**Pull request by mcvsubbu**

### CODE REVIEW

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java [hunk 1]
```diff
@@ -17,131 +17,206 @@
...
+     *
+     * An alternative is to not have multiple buffers, but just copy the values from one buffer to next as we
+     * increase the number of rows.
+     */
+    if (readers.size() == 1) {
+      readers.get(0).readIntValues(rows, 0, rowStartPos, rowSize, values, valuesStartPos);
+    }
+    int numRows = 0;
+    for (int rowIter = rowStartPos, valueIter = valuesStartPos; numRows < rowSize; numRows++, rowIter++, valueIter++) {
+      int row = rows[rowIter];
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java]: This avoids some div and modulo costs but not sure if it's more optimal (so optional suggestion)
prevBufferIndex = getBufferId(rows[0]);
startRowIndex = (prevBufferIndex) * numRowsInChunk
endIndex = startRowIndex + numRowsInChunk
for() {
   if rows[rowIter] >= startRowIndex && rows[rowIter] < endRowIndex {
       values[valueIter] = readers.get(prevBufferIndex).getInt(endRowIndex - rows[rowsIter]);
    } else {
      prevBufferIndex = getBufferId(rows[0]);
      startRowIndex = (prevBufferIndex) * numRowsInChunk
      endIndex = startRowIndex + numRowsInChunk
      values[valueIter] =  readers.get(prevBufferIndex).getInt(endRowIndex - rows[rowsIter]);
    }
}

}

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java [hunk 2]
```diff
@@ -17,131 +17,206 @@
...
+     * If we assume that the row IDs in 'rows' is sorted, then we can write logic to move values from one reader
+     * at a time, identifying the rows in sequence that belong to the same block. This logic is more complex, but may
+     * perform better in the sorted case.
+     *
+     * An alternative is to not have multiple buffers, but just copy the values from one buffer to next as we
+     * increase the number of rows.
+     */
+    if (readers.size() == 1) {
+      readers.get(0).readIntValues(rows, 0, rowStartPos, rowSize, values, valuesStartPos);
+    }
```
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java]: Good catch, thanks. Added.

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java (hunk 1 above)
Comment (mcvsubbu) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java]: makes sense. optimizes if the rows array is ordered.

**Comment by kishoreg** [comment]
Logic looks complicated and too many if statements during. This will impact Perf. Can we simplify this. We can chat tomorrow

File: pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java [hunk 3]
```diff
@@ -17,131 +17,223 @@
...
+    this.columnSizesInBytes = columnSizesInBytes;
+    addBuffer();
+  }
+
+  private void addBuffer() {
+    PinotDataBuffer buffer = PinotDataBuffer.allocateDirect(chunkSizeInBytes);
+    buffer.order(ByteOrder.nativeOrder());
+    buffers.add(buffer);
+    FixedByteSingleValueMultiColReader reader = new FixedByteSingleValueMultiColReader(buffer, numRowsPerChunk, new int[]{columnSizesInBytes});
+    currentWriter = new FixedByteSingleValueMultiColWriter(buffer, numRowsPerChunk, /*cols=*/1, new int[]{columnSizesInBytes});
```
Comment (antumbde) [inline pinot-core/src/main/java/com/linkedin/pinot/core/io/readerwriter/impl/FixedByteSingleColumnSingleValueReaderWriter.java]: if you use our new IDE then /*cols=*/ is not required :)

**Pull request merged**